pkg/steps/template.go

```diff
@@ -421,17 +421,57 @@ func createOrRestartPod(podClient coreclientset.PodInterface, pod *coreapi.Pod)
 	return created, nil
 }
 
-func waitForPodDeletion(podClient coreclientset.PodInterface, name string, uid types.UID) error {
+func waitForTeardownToComplete(podClient coreclientset.PodInterface, name string, uid types.UID) error {
 	timeout := 300
 	for i := 0; i < timeout; i += 2 {
-		pod, err := podClient.Get(name, meta.GetOptions{})
-		if errors.IsNotFound(err) {
-			return nil
+		pod, err := checkPodExists(podClient, name, uid)
+		if err != nil {
+			return err
+		}
+		if !podTeardownCompleted(pod) {
+			log.Printf("Waiting for pod %s to complete teardown ... (%ds/%d)", pod.GetName(), i, timeout)
+			time.Sleep(2 * time.Second)
+		} else {
+			break
 		}
+	}
+	return nil
+}
+
+func checkPodExists(podClient coreclientset.PodInterface, name string, uid types.UID) (*coreapi.Pod, error) {
+	pod, err := podClient.Get(name, meta.GetOptions{})
+	if errors.IsNotFound(err) {
+		return nil, errors.NewGone("pod was removed")
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve deleting pod: %v", err)
+	}
+	if pod.UID != uid {
+		return nil, errors.NewBadRequest("mismatching UID")
```

Comment:
Similarly ... a new pod with the same name but a different UUID just means teardown did complete?

Reply:
Different UUID means the existing pod is now gone and we no longer have to wait for teardown to complete, so its expected

Reply:
But you're returning an error that you then bubble up ..?